bellman/src/main.rs

```diff
@@ -0,0 +1,315 @@
+                            current_c_lc.lc = current_c_lc.lc + &next.lc;
+                            current_c_lc.debug = current_c_lc.debug + " + " + &next.debug;
+                            current_c_lc.result.add_assign(&next.result);
+                            advance = true;
+                        }
+                    }
+                    if advance {
+                        next_c_lc = read_next_constraint::<E>(&mut qap_c_file, &mut vars, true);
+                    } else {
+                        break;
+                    }
+                }
+
+                assert!(current_a_lc.index == current_b_lc.index);
+                assert!(current_a_lc.index == current_c_lc.index);
+                //println!("Constraint: {} * {} = {}", current_a_lc.debug, current_b_lc.debug, current_c_lc.debug);
+
+                cs.enforce(
+                    || current_a_lc.index.to_string(),
+                    |lc| lc + &current_a_lc.lc,
+                    |lc| lc + &current_b_lc.lc,
+                    |lc| lc + &current_c_lc.lc,
+                );
+
+                current_a_lc.result.mul_assign(&current_b_lc.result);
+                assert!(current_a_lc.result == current_c_lc.result, 
+                format!("constraint {} failed: {} * {} != {}", current_a_lc.index, current_a_lc.debug, current_b_lc.debug, current_c_lc.debug));
+
+                current_a_lc = empty_line(current_a_lc.index + 1);
+                current_b_lc = empty_line(current_b_lc.index + 1);
+                current_c_lc = empty_line(current_c_lc.index + 1);
+            }
+        }
+
+        /*
+         * For some reason, pepper can include variables in the witness that are not constrained.
+         * Bellmann asserts that this is not the case. We could clean up the witness from unbound
+         * variables and shift all constraints appropriately but adding dummy constraints seems to
+         * be the easier option for now.
+         */
+        for var in vars.iter() {
+            if !var.constrained {
+                println!("[Warning] Unconstrained variable {:?}", var.variable);
+                cs.enforce(
+                    || format!("unused var {:?}", var.variable),
+                    |lc| lc + var.variable,
+                    |lc| lc + CS::one(),
+                    |lc| lc + var.variable,
+                );
+            }
+        }
+
+        println!("  {} Done Reading", now.elapsed().as_secs());
+        Ok(())
+    }
+}
+
+fn print_usage(args: Vec<String>) {
+    println!("
+Program to create a zkSnark proof for an R1CS generated by the pepper framework.
+It requires the R1CS (A, B, C matrices) as well as assignments for primary and auxiliary input.
+
+The witnesses are consumed from a comma separated list of field elements.
+The matrices are consumed as files where each line corresponds to a nonzero entry in the corresponding QAP matrix. 
+E.g. a line `96 5 2` means the entry at 96th row, 5th column is 2.
```

Comment:
Are you sure this is not 5th row, 96th column?

Reply:
https://github.com/pepper-project/pequin/blob/master/fileformats.txt#L407-412

Reply:
They must be confused. That is simply not the case.

Reply:
Please read a little further.

https://github.com/pepper-project/pequin/blob/2fa79bfc434ce6f9d774c2897ce8da1aa35de9a7/fileformats.txt#L413-L416

For them,
rows <--> variables
columns <--> constraints

(which is backwards)

Reply:
Who says that rows=constraints and column=variable. It could equally be defined the other way around. I think what it should really say is: The 5th constraint contains the linear term (var_96 * 2)